File: 1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.c
char* dfs (char* prefix, int n, int n2, int* k) {
    if (n == 0)
        return prefix;
    for (char c = 'a'; c <= 'c'; c++) {
        int len = strlen (prefix);
        if (len > 0 && c == prefix [len - 1])
            continue;
        int count = (int) pow (2, n2 - len - 1);
        if (count >= *k) {
            prefix [len] = c;
            prefix [len + 1] = '\0';
            return dfs (prefix, n - 1, n2, k);
        } else
            *k -= count;
    }
    return "";
}

char* getHappyString (int n, int k) {
    int n2 = n;
    char* prefix = (char*) malloc ((n + 1) * sizeof (char));
    prefix [0] = '\0';
    return dfs (prefix, n, n2, &k);
}

File: 1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.py
class Solution:
    def getHappyString (self, n : int, k : int) -> str:
        def dfs (prefix, n, k):
            if n == 0:
                return prefix
            for c in 'abc':
                if prefix and c == prefix [-1]:
                    continue
                count = 2 ** (n2 - len (prefix) - 1)
                if count >= k:
                    return dfs (prefix + c, n - 1, k)
                else:
                    k -= count
            return ""
        
        n2 = n
        return dfs ("", n, k)